src/ComponentProvider.ts

```diff
@@ -0,0 +1,90 @@
+import type { Schema, Message, AsyncAPIDocument, ChannelParameter } from '@asyncapi/parser';
+
+/**
+ * This Singleton class will provide all sorts of data for optimizers.
+ *
+ * @public
+ */
+export class ComponentProvider {
+  messages = new Map<string, Message>();
+  schemas = new Map<string, Schema>();
+  parameters = new Map<string, ChannelParameter>();
+
+  private static instance: ComponentProvider;
+  private constructor(private document: AsyncAPIDocument) {
+    this.scanChannels();
+    this.scanComponents();
+  }
+  static getInstance(document: AsyncAPIDocument): ComponentProvider {
+    if (!ComponentProvider.instance) {
+      ComponentProvider.instance = new ComponentProvider(document);
+    }
+    return ComponentProvider.instance;
+  }
```

Comment:
It's not a good approach, because if you run in same time the `new Optimizer(doc)` two and more times then here you will always operate on the first instance. I think that `Optimizer class` should have the `componentProvider` field with instance of this class, so this class shouldn't be singleton.